src/activation_token/gtk3.rs

```diff
@@ -0,0 +1,60 @@
+use super::wayland::WaylandActivationToken;
+use gtk3::glib::translate::ToGlibPtr;
+use gtk3::prelude::*;
+use gtk3::{gdk, glib};
+use wayland_client::{
+    backend::ObjectId,
+    protocol::{__interfaces::WL_SURFACE_INTERFACE, wl_surface::WlSurface},
+    Proxy,
+};
+use wayland_protocols::xdg::activation::v1::client::xdg_activation_token_v1::XdgActivationTokenV1;
+
+#[derive(Debug)]
+pub struct Gtk3ActivationToken {
+    pub token: String,
+    inner: XdgActivationTokenV1,
+}
+
+impl Gtk3ActivationToken {
+    pub fn from_window(window: &impl glib::IsA<gdk::Window>) -> Option<Self> {
+        let display = window.as_ref().display();
+        match display.backend() {
+            gdk::Backend::Wayland => {
+                let cnx = wayland_client::Connection::connect_to_env().unwrap();
+                let wayland_window = window
+                    .as_ref()
+                    .downcast_ref::<gdk3wayland::WaylandWindow>()
+                    .unwrap();
+                let surface_id = unsafe {
+                    let ptr = gdk3wayland::ffi::gdk_wayland_window_get_wl_surface(
+                        wayland_window.to_glib_none().0,
+                    );
+
+                    ObjectId::from_ptr(&WL_SURFACE_INTERFACE, ptr as *mut _).unwrap()
+                };
+                let wl_surface = WlSurface::from_id(&cnx, surface_id).unwrap();
+                let wl = WaylandActivationToken::from_surface(&wl_surface).ok()?;
```

Comment:
Could use a custom constructor that takes a pointer and use it here & and in the gtk4 variant maybe?